text/0069-api-v2.md

```diff
@@ -0,0 +1,178 @@
+
+Note that in `API V2` mode, you can't specify CF in `RawClient`.
+
+Using TxnKV along with RawKV is safe in `API V2` mode.
+
+Listed below is the detailed compatibility matrix:
+
+|              | V1 Server | V2 Server |
+| ------------ | --------- | --------- |
+| V1 RawClient | Raw Data  | Forbidden |
+| V1 TxnClient | Txn Data  | TiDB Data |
+| V2 RawClient | Forbidden | Raw Data  |
+| V2 TxnClient | Forbidden | Txn Data  |
+
+### Keyspace
+
+When API V2 is enabled for the client, the user will be able to specify a keyspace for a session of `RawClient` or `TxnCient`. A keyspace is identified by a string, whose default value is 'default'.
+
+Keys in other keyspace will be invisible to the user.
+
+## Reference-level design
+
+### Key Encode
+
+Once the `API V2` is enabled, the key will be starting with either:
+
+1. `m` or `t`: TxnKV key. Used by TiDB.
+2. `x{keyspace prefix id}`: TxnKV key.
+3. `r{keyspace prefix id}`: RawKV key.
+4. `a{keyspace prefix id}`: RawKV atomic key, on which you can use `CompareAndSwap`.
+
+The `{keyspace prefix id}` is the [keyspace](https://github.com/tikv/rfcs/pull/39) prefix for seperating keys of different keyspace. It should be an vary-length integer whose highest bit of every byte denotes whether the next byte is still part of the integer. The client will fetch the prefix from PD by the keyspace name specified by the user when initializing the client, so it means that the keyspace prefix is valid during the session, in other words, the change on keyspace on PD will not take effect on running seesions.
+
+Note that in TxnKV, the key will be encoded by `Memory Comparable Encoding`. But since the `Memory Comparable Encoding` will not change the starting bytes and only add paddings, there won't be any overlap between RawKV and TxnKV.
+
+### RawKV Value Encode
+
+If the key is started with `r` or `a`, then the value can be either:
+
+1. `{0x0}{data}`
+2. `{0x1}{TTL expire timestamp}{data}`
+
+### PD
+
+Add a new http interface for adding, renaming, deleteing and querying the mapping from keyspace name to prefix id:
+
+```javascript
+// list all keyspaces
+GET /keyspaces
+[
+  {
+    name: "default",
+    prefix_id: "0"
+  }, 
+  {
+    name: "graghdb",
+    // a prefix id can be referred from mulitple keyspace name
+    prefix_id: "0"    
+  }, 
+  {
+    name: "redis",
+    prefix_id: "1"
+  }
+]
+
+// add new keyspace
+POST /keyspaces
+{
+  name: "mongodb",
+  // optional. if is null, PD will allocate an unused prefix id
+  prefix_id: "2"
+}
+
+// update a keyspace
+POST /keyspaces/{keyspace_name}
+// delete a keyspace
+DELETE /keyspaces/{keyspace_name}
+```
+
+The keyspace is stored in etcd and should have no limitation on the numbers.
+
+### TiKV
+
+In TiKV config file, add a new configuration `storage.api_version`. When enabled, `storage.enable_ttl` must also be enabled.
```

Comment:
what would happen if we enable `storage.api_version` and leave `storage.enable_ttl` disabled?

Reply:
The tikv will panic on start.